haskell: don't lose packageOverrides for native-bignum and integer-simple

cdepillabout opened the pull request:
###### Description of changes

This updates `haskell.packages.{native-bignum,integer-simple}.*` so that the new haskell packages added to `packageOverrides` are kept in the native-bignum and integer-simple package sets.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

cdepillabout committed:
haskell: don't lose packageOverrides for native-bignum

cdepillabout commented:
Here is a file that can be used to test this change:
`test.nix`:

```nix
let

  nixpkgs-src = ~/git/nixpkgs;

  my-overlay = final: prev: {
    haskell = prev.haskell // {
      packageOverrides =
        final.lib.composeExtensions
          (prev.haskell.packageOverrides or (_: _: {}))
          (hfinal: hprev: {
            my-new-package = hfinal.conduit;
          });
    };
  };

  pkgs = import nixpkgs-src { overlays = [my-overlay]; };
in
pkgs.pkgsStatic.haskellPackages.my-new-package
```

If you build this with the change from this PR, it successfully builds `conduit` through the `my-new-package` attribute:

```console
$ nix-build test.nix
...
/nix/store/l67n7i049jsvk9k3j87d0f5kmg7fmp4v-conduit-static-x86_64-unknown-linux-musl-1.3.4.2
```

However, if you're not using this PR, then you get an error about the `my-new-package` attribute not being able to be found (for instance, with commit d68508397445283e):

```console
$ nix-build test.nix
error: attribute 'my-new-package' missing

       at /home/illabout/temp/what012.nix:21:1:

           21| pkgs.pkgsStatic.haskellPackages.my-new-package
             | ^
```

cdepillabout commented on pkgs/top-level/haskell-packages.nix:
```diff
@@ -245,29 +245,43 @@ in {
-        integer-gmp = null;
-      };
-    });
+    native-bignum =
+      let
+        nativeBignumGhcNames = pkgs.lib.filter
+          (name: builtins.elem name nativeBignumIncludes)
+          (pkgs.lib.attrNames compiler);
+      in
+      pkgs.lib.genAttrs nativeBignumGhcNames
+        (name:
+          packages.${name}.override (oldAttrs: {
+            ghc = bh.compiler.native-bignum.${name};
+            buildHaskellPackages = bh.packages.native-bignum.${name};
+            overrides =
+              pkgs.lib.composeExtensions
+                (oldAttrs.overrides or (_: _: {}))
+                (_: _: { integer-gmp = null; });
+          })
+        );
```
I refactored this to make it a little easier to see exactly what is going on.

The only non-whitespace change is going from

```nix
      overrides = _self : _super : {
        integer-gmp = null;
      };
```

to

```nix
            overrides =
              pkgs.lib.composeExtensions
                (oldAttrs.overrides or (_: _: {}))
                (_: _: { integer-gmp = null; });
```

With this change, we make sure to keep any old overrides.

This is important because `package.${name}` is going to refer to one of the Haskell package sets.  For instance:

https://github.com/NixOS/nixpkgs/blob/af5b994d79e53709463ba4ff036f4ac50ebd3951/pkgs/top-level/haskell-packages.nix#L222-L226

You can see that this is defined with `callPackage`, which is defined in that same file:

https://github.com/NixOS/nixpkgs/blob/af5b994d79e53709463ba4ff036f4ac50ebd3951/pkgs/top-level/haskell-packages.nix#L31-L34

`callPackage` specifically defines `overrides` as `packageOverrides`.  The change in this PR makes sure not to lose this.

sternenseemann commented on pkgs/top-level/haskell-packages.nix:
```diff
@@ -245,29 +245,43 @@ in {
+                (oldAttrs.overrides or (_: _: {}))
+                (_: _: {
+                  integer-simple = null;
+                  integer-gmp = null;
```
This line can be dropped as it's `null` already in all `configuration*.nix` files.

cdepillabout replied:
Good call, I fixed this in 1be5cae549b.

sternenseemann commented on pkgs/top-level/haskell-packages.nix:
```diff
@@ -245,29 +245,43 @@ in {
+            overrides =
+              pkgs.lib.composeExtensions
+                (oldAttrs.overrides or (_: _: {}))
+                (_: _: { integer-gmp = null; });
```
This line can be dropped as it's `null` already in all `configuration*.nix` files, so we can drop this `composeExtensions` thing altogether.

With `native-bignum`, apparently, the alternative bignum backend is always included with GHC, even if GMP is used by default. With `integer-simple` GHCs, `integer-gmp` is always (like for `native-bignum`) included, but not `integer-simple`.

cdepillabout committed:
haskell: remove override of integer-gmp

It appears that integer-gmp is already set to null for all compilers,
so there is no need to explicitly set it to null in the integer-simple
and native-bignum package sets.
